### ISSUE TITLE
Implementation of 2d bicubic grid sampler

### ISSUE THREAD

**Issue by pomelyu** [pytorch/pytorch#44601]
## 🚀 Feature
support 2d bicubic interpolation in `nn.function.grid_sample` in https://github.com/pytorch/pytorch/issues/25039

## Pitch

Using the [2d bicubic convolution](https://en.wikipedia.org/wiki/Bicubic_interpolation) just as `torch.nn.functional.interpolate()` [here](https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/UpSampleBicubic2d.cpp).

## Additional context
I would like to implement this feature, and there are some questions
1. There are two cpu code about the grid_sample, `aten/src/Aten/native/GridSampler.cpp` and `aten/src/Aten/native/cpu/GridSamplerKernel.cpp`, and also the dispatch mechanism in https://github.com/pytorch/pytorch/blob/a309355be3d2c12dfdca17ee34a8259654c5b823/aten/src/ATen/native/GridSampler.cpp#L640
Are there any concern or recommend about the priority? I think I should first implement in `aten/src/Aten/native/GridSampler.cpp` and `aten/src/Aten/native/cuda/GridSampler.cu`. If success, try to implement in `aten/src/Aten/native/cpu/GridSamplerKernel.cpp` for AVX2

2. How could I test my implementation? There is `remap` function in openCV to test the forward correctness, but how about the backward? Use existed bicubic upsampling could cover some cases, but not general. Any suggestions would be appreciated

Thanks

cc @albanD @mruberry

**Comment by pomelyu** [comment]
Answer to myself. If the forward is implemented, the numerical method could be used to get the ground truth of backward gradient. i.e. forward a small amount dh, then get the output dH. dH/dh is the partial differential for the specific pixel output and pixel input

**Comment by mruberry** [comment]
Hey @pomelyu, 

I think we would accept a new interpolation mode for grid sample. It's tested in test_nn.py:

https://github.com/pytorch/pytorch/blob/b85568a54a9c60986235ad1e0cc5dffc71b9d5b1/test/test_nn.py#L6670

It looks like grid_sample also has a Python preamble in its function signature as well as the device generic dispatch and CPU and CUDA implementations. So you'd need to update all of these and add new tests validating the behavior.

I'd also look at which, if any, current grid sample interpolation modes are differentiable. If the other modes are differentiable I'd expect bicubic interpolation to be differentiable, too. 

Implementation question: do we have any helper methods for bicubic interpolation? Or is each implementation completely custom?

**Comment by pomelyu** [comment]
Hi @mruberry,

The other two interpolation, nearest and bilinear, and they are both differentiable, hence I'll also implement differentiable bicubic grid sampler.

There are bicubic interpolation in upsampling, hence I could use some functions from it.
